Add PyO3 mock data generation example

sanders41 opened the pull request:
@prrao87 can you fill in the performance part in the readme? Since you ran the benchmarks on your Mac for the other fake data generation I'd like to keep that consistent since I expect people will want to compare the numbers.

On my laptop, in release mode, generating 1000000 records took 0.991 seconds so I expect the numbers will be about the same as Rust alone.

prrao87 commented:
Hmm, this doesn't work on MacOS, as `libpython`can't be linked to as per [the docs](https://pyo3.rs/v0.20.2/building_and_distribution#macos), without some settings. The suggestion they suggest aren't working, so I guess I'll have to figure that out!

sanders41 commented:
Interesting, I thought I have used this same setup on Mac before. I wonder if this is specific to arm macs. 

Since it is using abi I wonder if we can remove the extension module.

prrao87 commented:
Couple of things. Did you mean remove `extension-module` and use just the `abi3` instead? This is what the docs say:

```toml
[dependencies]
pyo3 = { version = "0.20.2", features = ["abi3"] }
```

I notice that you've set it to `["abi3-py38"]` in Cargo.toml - is that because you're using Python 3.8?

Also, `maturin develop -r` works fine and that generates the wheel as expected. Can't we just pip install the wheel and use that instead?

prrao87 commented:
I added these lines to Cargo.toml as the PyO3 docs suggest:

```toml
[target.x86_64-apple-darwin]
rustflags = [
  "-C", "link-arg=-undefined",
  "-C", "link-arg=dynamic_lookup",
]

[target.aarch64-apple-darwin]
rustflags = [
  "-C", "link-arg=-undefined",
  "-C", "link-arg=dynamic_lookup",
]
```
I presume the bottom one is for arm macs? At least that's what I hope it is.

sanders41 commented:
`abi3-py38` builds a wheel that will work in 3.8+. Otherwise you have to build a wheel for every python version on every operating system.

I just tested and on my end removing the extension module from the `pyproject.toml` makes no difference so if that fixes things on your end let's do it. Or maybe we should just do it anyway? I notice no difference at all on my end.

I know I've used pyo3 on an M1 mac before. I'll see if I can find what I did there.

prrao87 commented:
It works after I remove the extension module!! 🎉 I just pushed the change that worked on my end.

prrao87 commented on src/pyo3_mock_data/pyproject.toml:
```diff
@@ -0,0 +1,42 @@
+
+[project]
+name = "prelude-parser"
+requires-python = ">=3.8"
```
My toml linter is complaining here, do you know why?

`{"name":"prelude-parser","requires-python":">=3.8"} is not valid under any of the given schemas`

prrao87 replied:
This was the even-better-toml extension on VS code, so I'm not sure if this is something invalid.

sanders41 replied:
I see why. I copied this from another project of mine and forgot to change the name :man_facepalming:

prrao87 replied:
Makes sense to have a toml linter also, I guess! So is prelude-parser a required dependency?

prrao87 commented:
Haha, a weird but good thing -- this code is actually running faster than my older mock data code. I get a timing of 1.1 sec on that one for generating 1M records, but in this PyO3 case I get ~0.9 sec. So a 15-20% speedup. Any thoughts as to why?

I made sure that both are running in release mode:

```sh
# Pure rust mock_data
cargo run -r -- 1000000

# PyO3 mock_data
python main.py -n 1000000
```

sanders41 commented:
I Fixed the names, see if that makes your linter happy.

> Haha, a weird but good thing -- this code is actually running faster than my older mock data code. I get a timing of 1.1 sec on that one for generating 1M records, but in this PyO3 case I get ~0.9 sec. So a 15-20% speedup. Any thoughts as to why?

The only thing I can think of that I changed is the error handling. I can't think of any reason that would make things faster though. If anything I would have thought it would make it slower.

prrao87 commented:
Nope, still exact same error:

`{"name":"pyo3-mock-data","requires-python":">=3.8"} is not valid under any of the given schemasEven Better TOML`

![image](https://github.com/thedataquarry/rustinpieces/assets/35005448/ef4c2d2a-fef8-49ac-86ac-509eead9996f)

Strange.

Also, could you run the old `mock_data` rust code a few times in release mode like I did? Just to see whether it's consistently slower. If so, might make sense to go back and we can fix the error handling in that one to be consistent. We definitely want both Rust code bases to be near-identical so that it's a fair comparison.

sanders41 commented:
> Nope, still exact same error:

I'm not sure what it is unhappy out here. It matches the documentation (Maturin and Setuptools both use [pep 621](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#python-requirement) https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#python-requirement.

> Also, could you run the old mock_data rust code a few times in release mode like I did?

I figured out the difference. `cargo run` adds some time for the build when in the Python version it is there is no build step before the run. When I run the Rust version like this I get an average time of 0.952s (make sure to run `cargo build -r` first).

```sh
../../../target/release/mock_data 1000000
```

prrao87 commented:
No worries about the toml, it's just a nuisance.

But I get what you mean now! I tested using your command after building the release version, and now the pure rust version is marginally faster (avg. of 0.89 sec on my end). I think we should be good to go!

sanders41 commented:
Sounds good. Yes, this number looks more like what I was expecting with pure Rust slightly faster than PyO3 and PyO3 way faster than Python.

prrao87 commented:
I'll go ahead and add my numbers to the benchmark then. Great stuff!

sanders41 commented:
Oops, I should have waited for that before merging 😆

prrao87 commented:
One more PR now worries 😄
